### PR TITLE
Geometry_Engine: Add next batch of unit tests

### LINKED ISSUE #2280: Geometry_Engine add next batch of unit tests
#### Definition of the test :
Next batch of tests for the following methods:
 `Compute.BooleanDifference`
 `Query.IsClockwise`
 `Modify.SplitAtPoints`
 `Query.PlaneIntersection`
 `Query.EndPoint`
 `Modify.Project`
 `Query.IsClosed`
 `Modify.Translate`
 `Query.StartPoint`
 `Query.Angle`
 `Create.Arc`
 `Query.IsPlanar`
 `Query.PointAtLength`
 `Compute.FitPlane`
 `Query.Clone`
 `Query.Degree`
 `Modify.Rotate`
 `Modify.Scale`
 `Compute.IntegrateRegion`
 `Query.Centre`
Based on the [discussion](https://github.com/BHoM/BHoM/discussions/1148).

### PR DESCRIPTION
**Pull request by rolyhudson**
### Issues addressed by this PR
Adding next unit tests prioritised by GitHub file changes.

Closes #2280 

### Test files
[Grasshopper scripts used for creation of tests](https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/BHoM_Engine/Geometry_Engine/%232309-NextBatchUnitTests?csf=1&web=1&e=yM3ckr).

### CODE REVIEW

**Comment by LMarkowski** [comment]
@BHoMBot check unit-tests

**Review by LMarkowski — CHANGES_REQUESTED** [review]
In general everything looks good. I've spotted a few tiny errors in scripts and a bunch of missing overloads. I'm not insisting on adding all of the missing methods right away, but as I started to note them I can as well share them here.

Errors:
- **Scale:** UT for `Mesh` is not plugged in.
- **Project:** UT for `PolySurface` is not plugged in.

Missing methods:
- **FitPlane:**  `Ellipse`, `Surface`, `ICurve`.
- **Rotate:** `Basis`, `Ellipse`, `PlanarSurface`, `IGeometry`, `ICurve`, `ISurface`.
- **Scale:** `Ellipse`, `PlanarSurface`, `IGeometry`, `ICurve`, `ISurface`.
- **Translate:** `Ellipse`, `PlanarSurface`, `IGeometry`, `ICurve`, `ISurface`.
- **Project:** `(Vector, Vector)`, `Ellipse`, `PlanarSurface`, `IGeometry`, `ICurve`, `ISurface`.
- **Clone:** `PlanarSurface`, `Basis`, `SurfaceTrim`, `IGeometry`, `ICurve`, `ISurface`.
- **EndPoint:** `Ellipse`, `ICurve`.
- **IsClockwise:** `ICurve`.
- **IsClosed:** `Ellipse`, `ICurve`.
- **PlaneIntersection:** `ICurve`.
- **PointAtLength:** `ICurve`.
- **StartPoint:** `Ellipse`, `ICurve`.
- **Centre:** `Polyline` (which is annotated as `ToBeRemoved` so I guess we don't need to create a special test for this one)

As one can see, besides few reoccurring classes (`Ellipse` and `PlanarSurface` 👀), what we lack the most (in the validation scripts) are the tests for interface methods. The question is whether we want to have them tested in the UTs. I think eventually we should. Most of them are just pointers to other methods, but I remember the cases when they do more than that.

**Comment by rolyhudson** [comment]
@BHoMBot check unit-tests

**Comment by FraserGreenroyd** [comment]
@BHoMBot check unit tests

**Comment by rolyhudson** [comment]
@BHoMBot check unit tests

**Comment by rolyhudson** [comment]
@BHoMBot check unit tests

**Comment by rolyhudson** [comment]
@BHoMBot check all

**Comment by rolyhudson** [comment]
@LMarkowski - addressed all your suggestions thanks for the review. 
Unit test check is failing due to errors from an unrelated unit test and warning on deserialization on an existing unit test. Probably I should request dispensation (maybe raise issues to fix the error and warning) but hesitating as I am uncertain of what protocol should be in this situation. @FraserGreenroyd what is your view on this scenario?

**Comment by FraserGreenroyd** [comment]
> @LMarkowski - addressed all your suggestions thanks for the review.
> Unit test check is failing due to errors from an unrelated unit test and warning on deserialization on an existing unit test. Probably I should request dispensation (maybe raise issues to fix the error and warning) but hesitating as I am uncertain of what protocol should be in this situation. @FraserGreenroyd what is your view on this scenario?

All unit tests run because changes to Geometry Engine could impact other engines, so normally I would say we would need to investigate any failures and get info to make decisions on. But in this particular scenario, this PR only adds unit tests rather than modifying core code, so I would say we need to investigate the Environment Unit Test and probably resolve it as it appears that unit test is expecting an object rather than an IRegion, which could be because it was made in the earlier days of the unit test method building approach.

The deserialisation of the Geometry Engine unit test though should probably be fixed in this PR because it relates to the unit tests of the engine this PR is modifying. But I will accept an issue being raised to investigate it later if you prefer.

To summarise - dispensation will be granted for Environment unit test, and can be granted for Geometry one if an issue is raised.

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a CI/CD instruction. I am authorising dispensation to be granted on check ref. 1953959901

**Comment by rolyhudson** [comment]
@BHoMBot check unit tests

**Comment by peterjamesnugent** [comment]
@rolyhudson might be outside the scope of this PR but there's a bug in the `IsPlanar` method where you can get it to always return true as long as you don't specify a tolerance with a `PlanarSurface` - not sure if it affects any of the unit tests you've just written.

**Comment by rolyhudson** [comment]
@peterjamesnugent Cheers for the note. It is a little out of scope for this PR as it is not yet covering the `IsPlanar` method. We probably should also ask the question why are we able to create non-planar `PlanarSurface`s in the first place? 😄

**Comment by rolyhudson** [comment]
@BHoMBot check unit tests

**Comment by rolyhudson** [comment]
@BHoMBot check unit tests

**Comment by rolyhudson** [comment]
@FraserGreenroyd @LMarkowski I have fixed the issue on the Grid UT. 
In the original Grid UT version inputs had somehow not been provided in the correct order and one as an empty input. My changes have fixed that and made no other changes to the original input data.

Assuming the previously granted dispensation for the fail on the MapRegions UT still stands (or do I need to re-request?) this is ready for a final review when you have time @LMarkowski. Cheers both.

**Comment by FraserGreenroyd** [comment]
> Assuming the previously granted dispensation for the fail on the MapRegions UT still stands (or do I need to re-request?) this is ready for a final review when you have time @LMarkowski. Cheers both.

It will need to be re-requested once ready for merge, but currently unit-tests is not a required check so even if it's a red cross we can still merge as we have trace of why we're merging with a red cross on this PR :smile:

**Review by LMarkowski — CHANGES_REQUESTED** [review]
Many thanks for adding all the missing tests! Unfortunately I found two more that I didn't notice on the first review. Sorry for that. 😓 

Missing methods:
- **Translate:**  `NurbsCurve`,
- **IsClosed:** `NurbsSurface`.

**Comment by rolyhudson** [comment]
Cheers @LMarkowski added those missing methods now.

**Comment by rolyhudson** [comment]
@BHoMBot you need a hyphen?

**Review by LMarkowski — APPROVED** [review]
Perfect!

**Comment by rolyhudson** [comment]
@BHoMBot check installer

**Comment by rolyhudson** [comment]
@BHoMBot check copyright-compliance

**Comment by rolyhudson** [comment]
@BHoMBot check project-compliance

**Comment by rolyhudson** [comment]
@BHoMBot check-serialisation

**Comment by rolyhudson** [comment]
@BHoMBot check-versioning

**Comment by rolyhudson** [comment]
@BHoMBot check core

**Comment by rolyhudson** [comment]
@BHoMBot check installer

**Comment by rolyhudson** [comment]
@FraserGreenroyd good to merge here now. Thanks

**Comment by FraserGreenroyd** [comment]
@BHoMBot check versioning

**Comment by FraserGreenroyd** [comment]
@BHoMBot check serialisation

**Comment by FraserGreenroyd** [comment]
@BHoMBot check serialisation

**Comment by FraserGreenroyd** [comment]
@BHoMBot check serialisation

**Comment by FraserGreenroyd** [comment]
@BHoMBot check serialisation
